src/deluge/model/song/clip_iterators.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <cstdint>
+#include <optional>
+
+#include "definitions_cxx.hpp"
+
+#ifdef IN_UNIT_TESTS
+#include "clip_iterator_mocks.h"
+#else
+#include "deluge.h"
+#include "model/clip/clip.h"
+#include "model/clip/clip_array.h"
+#include "model/song/song.h"
+#endif
+
+// This is an incomplete iterator, missing operator*() and operator->(): the templated iterators inherit
+// from this and implement just those for minimal template bloat.
+class ClipIteratorBase {
+public:
+	__attribute__((noinline))
+	ClipIteratorBase(ClipArray* array_, uint32_t index_, ClipArray* nextArray_, std::optional<ClipType> clipType_);
+	__attribute__((noinline)) ClipIteratorBase& operator++();
+	__attribute__((noinline)) ClipIteratorBase operator++(int);
+	/** Deletes the clip pointed to by the iterator and advances it. */
+	__attribute__((noinline)) void deleteClip(InstrumentRemoval instrumentRemoval);
+	__attribute__((noinline)) friend bool operator==(const ClipIteratorBase& a, const ClipIteratorBase& b) {
+		return (a.array == b.array) && (a.index == b.index) && (a.clipType == b.clipType);
+	}
+	__attribute__((noinline)) friend bool operator!=(const ClipIteratorBase& a, const ClipIteratorBase& b) {
```

Comment:
We've generally been using the official C++ attribute syntax, so these should be `[[gnu::noinline]]` rather than `__attribute__`